src/renderers/shared/fiber/ReactFiberUpdateQueue.js

```diff
@@ -367,14 +377,18 @@ exports.getPendingPriority = getPendingPriority;
 function addTopLevelUpdate(
   fiber : Fiber,
   partialState : PartialState<any, any>,
+  callback : Callback | null,
   priorityLevel : PriorityLevel
 ) : void {
-  const isTopLevelUnmount = partialState === null;
+  const isTopLevelUnmount = Boolean(
+    partialState &&
+    partialState.element === null
+  );
```

Comment:
Your suggested change is wrong because

```js
Object.assign({}, { element: element }, null);
```

merely copies the previous state, whereas an unmount sets `element` to null.